File: src/main/java/me/ialistannen/toydnarnaasconverter/util/EnumUtil.java
package me.ialistannen.toydnarnaasconverter.util;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

/**
 * A Util to help with enums.
 */
public class EnumUtil {

  /**
   * @param enumClass The class of the enum
   * @param transformation The transformation to apply
   * @param <T> The enum class
   * @param <V> The class of the transformed value
   * @return A reverse-mapping of the enum
   */
  public static <T extends Enum<T>, V> Map<V, T> getReverseMultiMapping(Class<T> enumClass,
      Function<T, Collection<V>> transformation) {

    Map<V, T> result = new HashMap<>();

    for (T t : enumClass.getEnumConstants()) {
      for (V v : transformation.apply(t)) {
        result.put(v, t);
      }
    }

    return result;
  }
}

File: src/main/kotlin/me/ialistannen/toydnarnaasconverter/util/EnumUtil.kt
package me.ialistannen.toydnarnaasconverter.util

/**
 * @param enumClass The class of the enum
 * @param transformation The transformation to apply
 * @return A reverse-mapping of the enum
 */
fun <T : Enum<T>, V> getReverseMultiMapping(enumClass: Class<T>,
                                            transformation: (T) -> Collection<V>): Map<V, T> {

    val result = HashMap<V, T>()

    for (t in enumClass.enumConstants) {
        for (v in transformation.invoke(t)) {
            result.put(v, t)
        }
    }

    return result
}
